File: domain/src/main/java/org/cryptomator/domain/usecases/cloud/CancelAwareDataSource.java
package org.cryptomator.domain.usecases.cloud;

import android.content.Context;

import org.cryptomator.domain.exception.CancellationException;
import org.cryptomator.util.Optional;

import java.io.IOException;
import java.io.InputStream;

public class CancelAwareDataSource implements DataSource {

	private final DataSource delegate;
	private final Flag cancelled;

	private CancelAwareDataSource(DataSource delegate, Flag cancelled) {
		this.delegate = delegate;
		this.cancelled = cancelled;
	}

	public static CancelAwareDataSource wrap(DataSource delegate, Flag cancelled) {
		return new CancelAwareDataSource(delegate, cancelled);
	}

	@Override
	public Optional<Long> size(Context context) {
		if (cancelled.get()) {
			throw new CancellationException();
		}
		return delegate.size(context);
	}

	@Override
	public InputStream open(Context context) throws IOException {
		if (cancelled.get()) {
			throw new CancellationException();
		}
		return CancelAwareInputStream.wrap(delegate.open(context), cancelled);
	}

	public CancelAwareDataSource decorate(DataSource delegate) {
		return new CancelAwareDataSource(delegate, cancelled);
	}

	@Override
	public void close() throws IOException {
		delegate.close();
	}
}

File: domain/src/main/java/org/cryptomator/domain/usecases/cloud/CancelAwareDataSource.kt
package org.cryptomator.domain.usecases.cloud

import android.content.Context
import org.cryptomator.domain.exception.CancellationException
import java.io.IOException
import java.io.InputStream
import java.util.Date

class CancelAwareDataSource private constructor(private val delegate: DataSource, private val cancelled: Flag) : DataSource {

	override fun modifiedDate(context: Context): Date? {
		if (cancelled.get()) {
			throw CancellationException()
		}
		return delegate.modifiedDate(context)
	}

	override fun size(context: Context): Long? {
		if (cancelled.get()) {
			throw CancellationException()
		}
		return delegate.size(context)
	}

	@Throws(IOException::class)
	override fun open(context: Context): InputStream {
		if (cancelled.get()) {
			throw CancellationException()
		}
		return CancelAwareInputStream.wrap(delegate.open(context), cancelled)
	}

	override fun decorate(delegate: DataSource): CancelAwareDataSource {
		return CancelAwareDataSource(delegate, cancelled)
	}

	@Throws(IOException::class)
	override fun close() {
		delegate.close()
	}

	companion object {

		@JvmStatic
		fun wrap(delegate: DataSource, cancelled: Flag): CancelAwareDataSource {
			return CancelAwareDataSource(delegate, cancelled)
		}
	}
}
